provider/file/file_test.go

```diff
@@ -0,0 +1,196 @@
+package file
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"path"
+	"sync"
+	"testing"
+
+	"github.com/containous/traefik/safe"
+	"github.com/containous/traefik/types"
+	"github.com/stretchr/testify/assert"
+)
+
+const (
+	frontends = `
```

Comment:
I think, you can reduce test definitions like that:

<details>
<summary>file_test.go</summary>

```golang
package file

import (
	"context"
	"fmt"
	"io/ioutil"
	"os"
	"path"
	"testing"
	"time"

	"github.com/containous/traefik/provider"
	"github.com/containous/traefik/safe"
	"github.com/containous/traefik/types"
	"github.com/stretchr/testify/assert"
)

func TestProvideSingleFile(t *testing.T) {
	tempDir := createTempDir(t, "testfile")
	defer os.RemoveAll(tempDir)

	expectedNumFrontends := 2
	expectedNumBackends := 2

	tempFile := createFile(t,
		tempDir, "simple.toml",
		createFrontendConfiguration(expectedNumFrontends),
		createBackendConfiguration(expectedNumBackends))

	configurationChan, signal := createConfigurationRoutine(t, &expectedNumFrontends, &expectedNumBackends)

	fileProvider := &Provider{
		BaseProvider: provider.BaseProvider{
			Filename: tempFile.Name(),
			Watch:    true,
		},
	}
	fileProvider.Provide(configurationChan, safe.NewPool(context.Background()), nil)

	// Wait for initial message to be tested
	waitForSignal(t, signal, 2*time.Second, "initial config")

	// Now test again with single frontend and backend
	expectedNumFrontends = 1
	expectedNumBackends = 1

	tempFile = createFile(t,
		tempDir, "simple.toml",
		createFrontendConfiguration(expectedNumFrontends),
		createBackendConfiguration(expectedNumBackends))

	waitForSignal(t, signal, 2*time.Second, "single frontend and backend")
}

func TestProvideDirectory(t *testing.T) {
	tempDir := createTempDir(t, "testdir")
	defer os.RemoveAll(tempDir)

	expectedNumFrontends := 2
	expectedNumBackends := 2

	tempFile1 := createRandomFile(t, tempDir, createFrontendConfiguration(expectedNumFrontends))
	tempFile2 := createRandomFile(t, tempDir, createBackendConfiguration(expectedNumBackends))

	configurationChan, signal := createConfigurationRoutine(t, &expectedNumFrontends, &expectedNumBackends)

	fileProvider := &Provider{
		BaseProvider: provider.BaseProvider{
			Watch: true,
		},
		Directory: tempDir,
	}
	fileProvider.Provide(configurationChan, safe.NewPool(context.Background()), nil)

	// Wait for initial config message to be tested
	waitForSignal(t, signal, 2*time.Second, "initial config")

	// Now remove the backends file
	expectedNumFrontends = 2
	expectedNumBackends = 0
	os.Remove(tempFile2.Name())
	waitForSignal(t, signal, 2*time.Second, "remove the backends file")

	// Now remove the frontends file
	expectedNumFrontends = 0
	expectedNumBackends = 0
	os.Remove(tempFile1.Name())
	waitForSignal(t, signal, 2*time.Second, "remove the frontends file")
}

func createConfigurationRoutine(t *testing.T, expectedNumFrontends *int, expectedNumBackends *int) (chan types.ConfigMessage, chan interface{}) {
	configurationChan := make(chan types.ConfigMessage)
	signal := make(chan interface{})

	go func() {
		for {
			data := <-configurationChan
			assert.Equal(t, "file", data.ProviderName)
			assert.Len(t, data.Configuration.Frontends, *expectedNumFrontends)
			assert.Len(t, data.Configuration.Backends, *expectedNumBackends)
			signal <- nil
		}
	}()

	return configurationChan, signal
}

func waitForSignal(t *testing.T, signal chan interface{}, timeout time.Duration, caseName string) {
	timer := time.NewTimer(timeout)
	defer timer.Stop()

	select {
	case <-signal:

	case <-timer.C:
		t.Fatalf("Timed out waiting for assertions to be tested: %s", caseName)
	}
}

// createRandomFile Helper
func createRandomFile(t *testing.T, tempDir string, contents ...string) *os.File {
	return createFile(t, tempDir, fmt.Sprintf("temp%d.toml", time.Now().UnixNano()), contents...)
}

// createFile Helper
func createFile(t *testing.T, tempDir string, name string, contents ...string) *os.File {
	fileName := path.Join(tempDir, name)

	tempFile, err := os.Create(fileName)
	if err != nil {
		t.Fatal(err)
	}

	for _, content := range contents {
		_, err := tempFile.WriteString(content)
		if err != nil {
			t.Fatal(err)
		}
	}

	err = tempFile.Close()
	if err != nil {
		t.Fatal(err)
	}

	return tempFile
}

// createTempDir Helper
func createTempDir(t *testing.T, dir string) string {
	d, err := ioutil.TempDir("", dir)
	if err != nil {
		t.Fatal(err)
	}
	return d
}

// createFrontendConfiguration Helper
func createFrontendConfiguration(n int) string {
	conf := "[frontends]\n"
	for i := 1; i <= n; i++ {
		conf += fmt.Sprintf(`  [frontends.frontend%[1]d]
  backend = "backend%[1]d"
`, i)
	}
	return conf
}

// createBackendConfiguration Helper
func createBackendConfiguration(n int) string {
	conf := "[backends]\n"
	for i := 1; i <= n; i++ {
		conf += fmt.Sprintf(`  [backends.backend%[1]d]
    [backends.backend%[1]d.servers.server1]
    url = "http://172.17.0.%[1]d:80"
`, i)
	}
	return conf
}

```
</details>